src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -108,6 +108,41 @@ record = Marshal(item);
             }
         }
 
+        public override async Task<DValue<BooleanValue>> RemoveAsync(IEnumerable<FormulaValue> recordsToRemove, bool all = false)
+        {
+            var ret = false;
+
+            if (_sourceList == null || _sourceList.IsReadOnly)
+            {
+                return await base.RemoveAsync(recordsToRemove, all);
+            }
+
+            // !JYL! Needs improvement
+            foreach (RecordValue record in recordsToRemove)
+            {
+                while (true)
+                {
+                    var actual = Find(record);
+
+                    if (actual != null)
+                    {
+                        _sourceList.Remove(MarshalInverse(actual));
```

Comment:
> _sourceList.Remove(MarshalInverse(actual)); [](http://example.com/codeflow?start=23&length=44)

This may not work reliably. 

MarshalInverse can return a new instance of T, which may not Object.Equals match to the T in the list. 


We may need to iterate on _enumerator directly. 

var deleteList = new List<T>();
foreach(var item in _enumerator)
{
  if (Matches(item, record) {
     deleteList.Add(item); 
  }
}

foreach(var delete in deleteList)
{
  // delete is the same object instance as what was in the list. 
  _sourceList.Remove(delete);
}